src/util/taproot.rs

```diff
@@ -82,12 +93,804 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
     doc="Taproot-tagged hash for the taproot signature hash", true
 );
 
+impl TapTweakHash {
+
+    /// Create a new BIP341 [`TapTweakHash`] from key and tweak
+    /// Produces H_taptweak(P||R) where P is internal key and R is the merkle root
+    pub fn from_key_and_tweak(
+        internal_key: schnorr::PublicKey,
+        merkle_root: Option<TapBranchHash>,
+    ) -> TapTweakHash {
+        let mut eng = TapTweakHash::engine();
+        // always hash the key
+        eng.input(&internal_key.serialize());
+        if let Some(h) = merkle_root {
+            eng.input(&h);
+        } else {
+            // nothing to hash
+        }
```

Comment:
nit: 
```suggestion
        `merkle_root.map(|h| eng.input(h))`
```

Reply:
It might be a matter of personal taste, but I don't like using `map` when it mutates the external environmental variables (eng in this case). I see map as converting Option<T> -> Option<U>. 

In cases where there is an environmental mutation, I prefer using loops/explicit if let over functional methods like `iter(), map()`. In other words, I only like functional methods when there are no side effects.